### PR TITLE
Add support for std::array<T, N>

### LINKED ISSUE #5086: Add InputArray and OutputArray overloads for std::array
Currently, there are overloads for InputArray and OutputArray that take an std::vector:

```
template<typename _Tp> _InputOutputArray(std::vector<Mat_<_Tp> >& vec);
```

It would be great to also have overloads that take an std::array, to make the following code work:

```
cv::Mat image;
std::vector<cv::Mat> output(3, cv::Mat());
cv::split(image, output); // This works

std::array<cv::Mat, 3> output_array;
cv::split(image, output_array); // This currently doesn't work yet
```

### PR DESCRIPTION
**Pull request by arnaudbrejeon**
resolves #5086

Partially adds support for std::array from request "Add InputArray and OutputArray overloads for std::array #5086"

* `CV_CXX_STD_ARRAY` is defined to know if `std::array` is supported by the compiler
* New constructors for `InputArray`, `OutputArray`, `InputOutputArray`, and `Mat` that take `std::array` and `std::array<Mat>` as input.
* Add tests (similar to `std::vector` tests)

### CODE REVIEW

File: modules/core/src/copy.cpp [hunk 1]
```diff
@@ -292,7 +292,7 @@ void Mat::copyTo( OutputArray _dst ) const
         {
             // For some cases (with vector) dst.size != src.size, so force to column-based form
             // It prevents memory corruption in case of column-based src
-            if (_dst.isVector())
+            if (_dst.isVector() || _dst.isArray())
```
Comment (alalek) [inline modules/core/src/copy.cpp]: Probably we should avoid introduction of new "isArray()" method and merge it with `isVector()`.

Comment (arnaudbrejeon) [inline modules/core/src/copy.cpp]: Good idea. I will update the commits to take this into account

**Comment by arnaudbrejeon** [comment]
I've removed isArray and added support for std::array<cv::Mat>

File: modules/core/include/opencv2/core/mat.hpp [hunk 2]
```diff
@@ -2281,6 +2323,12 @@ class CV_EXPORTS UMat
     UMat(const UMat& m, const std::vector<Range>& ranges);
     //! builds matrix from std::vector with or without copying the data
     template<typename _Tp> explicit UMat(const std::vector<_Tp>& vec, bool copyData=false);
+
+#ifdef CV_CXX_STD_ARRAY
+    //! builds matrix from std::array with or without copying the data
+    template<typename _Tp, std::size_t _N> explicit UMat(const std::array<_Tp, _N>& vec, bool copyData=false);
+#endif
```
Comment (alalek) [inline modules/core/include/opencv2/core/mat.hpp]: Please remove this overload.
UMat usage is highly integrated with OpenCL, which usually requires large enough amount of data (there is significant overhead for small arrays). In general, we should not allow users/developers to write a slow code.

**Comment by alalek** [comment]
Well done! :+1:

**Pull request merged**